File: SubpopulationCMSA.m
classdef SubpopulationCMSA < Subpopulation  % inherits the Subpopulation class
    % ----- Additional attributes of subpopulations in the core search is CMSA -----"""
    properties
        elite; % elite solutions
    end
    methods
        function subpop=SubpopulationCMSA(center,smean,stretch,popSize) % Constructor
            subpop@Subpopulation(center,smean,stretch,popSize); % initialize the properties of the parent class
            subpop.elite=Elite(numel(stretch)); % repetition of the subpopulation mean for the elite solutions of the subpopulation  
        end
        
        function sample_solutions(subpop,restart,archive,process,opt,problem)  
            % *** Sample new popualtion members and evaluate them ***  
            tempRedRatio=1; % For temoprary reduction of the taboo region if too many samples are rejected
            solNo=1; % solution number: counter for created taboo acceptable solutions

            % Preallocate values
            subpop.samples=Sampling(process.subpopSize,problem.dim); 


            % Sampling stage: generate taboo-acceptable solutions
            while solNo<=process.subpopSize
                %print(tempRedRatio)
                subpop.samples.s(solNo)=subpop.mutProfile.smean*exp(randn*process.coreSpecStrPar.tauSigma); % calculate individual step sizes
                subpop.samples.Z(solNo,:)=(subpop.mutProfile.rotMat*(subpop.mutProfile.stretch.*randn(1,problem.dim))')'; % vector Z

                
                subpop.samples.X(solNo,:)=subpop.center+subpop.samples.s(solNo)*subpop.samples.Z(solNo,:); % sampled solution
                subpop.repair_infeas(solNo,opt,problem); % relocate to the search range if outside

                %  now check if this sampled solution is taboo acceptable 
                acceptIt=subpop.is_taboo_acceptable(subpop.samples.X(solNo,:),tempRedRatio,opt); % check if the samples solution is taboo acceptable 
                if acceptIt % if it is taboo acceptable
                    solNo=solNo+1;
                else
                    tempRedRatio=tempRedRatio*opt.niching.redCoeff; % temperoray reductions of taboo region sizes if too many attempts failed consecutively
                end
            end  
        end
        
        
        
        function recombine(subpop,restart,archive,process,opt,problem) 
            % -------- Perform recombination when the core search is elitist CMSA -----------
            oldCenter=subpop.center;
            ind=subpop.samples.argsortWithElite;  % use sorted indexes from elite                    

            % update the center  
            subpop.center=process.recWeights*subpop.samples.X(ind(1:process.mu),:); % update the center of the subpopulation
            % Relocate Xmean to the closest point in the search range
            subpop.center=max([subpop.center;problem.lowBound]);
            subpop.center=min([subpop.center;problem.upBound]);

            % --------- update the best solujtion of the subpopulation ------------
            subpop.bestSol=subpop.samples.X(ind(1),:);
            subpop.bestVal=subpop.samples.f(ind(1)); 

            % ----------- update the global step size -------------
            correctionTerm=(geomean(subpop.samples.s)/subpop.mutProfile.smean)^opt.coreSearch.sigmaUpdateBiasImp; % Correction for rejection of some generated solutions (for unsymmetric distribution os individul step sizes)
            subpop.mutProfile.smean=prod(subpop.samples.s(ind(1:process.mu)).^process.recWeights) / correctionTerm;  % suggested  global step size by recombination

            % ---------- Update the covariance matrix ----------- 
            suggC=zeros(problem.dim,problem.dim); %suggested C based on the new solution evalautions % 
            for parNo=1:process.mu
                Z=(subpop.samples.X(ind(parNo),:)-oldCenter)/subpop.samples.s(ind(parNo)); % Z wrt the center in the just finished iteration
                suggC=suggC+process.recWeights(parNo)*(Z'*Z);
            end
            cc=1/process.coreSpecStrPar.tauCov; % learning rate for C 
            newC=(1-cc)*subpop.mutProfile.C+ cc*suggC; % update C
            subpop.mutProfile.C=0.5*(newC+newC'); % enforce symmetry
            [subpop.mutProfile.rotMat,tmp]=eig(subpop.mutProfile.C); % perform eigen decomposition to calculate rotMat and stretch
            subpop.mutProfile.stretch=sqrt(diag(tmp))';
            clear tmp
            subpop.mutProfile.Cinv= subpop.mutProfile.rotMat*diag(subpop.mutProfile.stretch.^(-2))*subpop.mutProfile.rotMat'; % inverce of the covariance matrix

            % ------------- Update elite solutions ----------------
            if process.coreSpecStrPar.numElt>0
                [~,surviveInd]=sort(subpop.samples.f);  
                limit1=sum(subpop.samples.isFeas)+numel(subpop.elite.val); % only feasible solutions  
                limit2=process.coreSpecStrPar.numElt; % another upper limit for the number of elite solutions
                actEltNum=floor(0.5+min(limit1,limit2)); % the number of elite solutions for the next iteration
                subpop.elite.sol=subpop.samples.X(surviveInd(1:actEltNum),:); 
                subpop.elite.val=subpop.samples.f(surviveInd(1:actEltNum)); 
                subpop.elite.Z=subpop.samples.Z(surviveInd(1:actEltNum),:); 
                subpop.elite.s=subpop.samples.s(surviveInd(1:actEltNum));
                subpop.elite.wasRepaired=subpop.samples.wasRepaired(surviveInd(1:actEltNum));
            end
        end
        function select(subpop,restart,archive,process,opt,problem)   
            % ----------- Perform selection -------------------  
            [~,subpop.samples.argsortNoElite]=sort(subpop.samples.f); % indexes of best non-elite samples

            % --- Append acceptable elite solutions to the seletion pool if the core search uses elitism and the elites are taboo acceptable ---- 
            if strcmp(opt.coreSearch.algorithm, 'CMSA') && (process.coreSpecStrPar.numElt>0) && numel(subpop.elite.val)>0 % if elitism is activated and there are elite solutions from the previous restart
                appendIt=ones(1,numel(subpop.elite.val)); % Preallocation
                % -------- only append elite solutions that are taboo acceptable: Find their indexes -----------
                for eltNo=1:numel(subpop.elite.val)
                    appendIt(eltNo)=subpop.is_taboo_acceptable(subpop.elite.sol(eltNo,:),1,opt); % only taboo elite solutions can be added (no temporary shrinkage of taboo regions is allowed in this case)
                end
                appendIt=find(appendIt);
                % ------ append the approved elite solutions to the selection pool -------
                subpop.samples.X=[subpop.samples.X;subpop.elite.sol(appendIt,:)];
                subpop.samples.Z=[subpop.samples.Z;subpop.elite.Z(appendIt,:)];
                subpop.samples.s=[subpop.samples.s,subpop.elite.s(appendIt)];
                subpop.samples.f=[subpop.samples.f,subpop.elite.val(appendIt)];
                subpop.samples.wasRepaired=[subpop.samples.wasRepaired,subpop.elite.wasRepaired(appendIt)];
            end  % if

            % ----- For performing elite selection 
            [~,subpop.samples.argsortWithElite]=sort(subpop.samples.f); % indices of best solutions (sorted)
        end % function

    end
end
        

File: subpopulation_cmsa.py
"""
SubpopulationCMSA - CMSA variant of Subpopulation with elite solutions.

Written by Ali Ahrari (aliahrari1983@gmail.com)
Translated to Python from MATLAB
"""

import numpy as np
from subpopulation import Subpopulation
from mutation_sampling import Sampling, Elite


class SubpopulationCMSA(Subpopulation):
    """Subpopulation using CMSA (Covariance Matrix Self-Adaptation) evolution strategy."""
    
    def __init__(self, center, smean, stretch, popSize):
        """Initialize CMSA subpopulation."""
        super().__init__(center, smean, stretch, popSize)
        self.elite = Elite(len(stretch))
    
    def sample_solutions(self, restart, archive, process, opt, problem):
        """Sample new population members and evaluate them."""
        tempRedRatio = 1.0  # For temporary reduction of taboo region if too many samples are rejected
        solNo = 0  # Solution counter
        
        # Preallocate
        self.samples = Sampling(process.subpopSize, problem.dim)
        
        # Sampling stage: generate taboo-acceptable solutions
        while solNo < process.subpopSize:
            # Calculate individual step sizes
            self.samples.s[solNo] = self.mutProfile.smean * np.exp(np.random.randn() * process.coreSpecStrPar.tauSigma)
            # Clamp tiny/invalid step sizes
            if not np.isfinite(self.samples.s[solNo]) or self.samples.s[solNo] < 1e-12:
                self.samples.s[solNo] = 1e-12
            
            # Generate vector Z
            self.samples.Z[solNo, :] = (self.mutProfile.rotMat @ 
                                        (self.mutProfile.stretch * np.random.randn(problem.dim)))
            
            # Sampled solution
            self.samples.X[solNo, :] = self.center + self.samples.s[solNo] * self.samples.Z[solNo, :]
            
            # Relocate to search range if outside
            self.repair_infeas(solNo, opt, problem)
            
            # Check if this sampled solution is taboo acceptable
            acceptIt = self.is_taboo_acceptable(self.samples.X[solNo, :], tempRedRatio, opt)
            
            if acceptIt:
                solNo += 1
            else:
                tempRedRatio *= opt.niching.redCoeff  # Temporary reduction of taboo region sizes
    
    def recombine(self, restart, archive, process, opt, problem):
        """Perform recombination when the core search is elitist CMSA."""
        oldCenter = self.center.copy()
        ind = self.samples.argsortWithElite
        
        # Update the center
        self.center = process.recWeights @ self.samples.X[ind[:process.mu], :]
        
        # Relocate center to the closest point in the search range
        self.center = np.maximum(self.center, problem.lowBound)
        self.center = np.minimum(self.center, problem.upBound)
        
        # Update the best solution of the subpopulation
        self.bestSol = self.samples.X[ind[0], :]
        self.bestVal = self.samples.f[ind[0]]
        
        # Update the global step size
        geomean_s = np.exp(np.mean(np.log(self.samples.s)))
        correctionTerm = (geomean_s / self.mutProfile.smean) ** opt.coreSearch.sigmaUpdateBiasImp
        
        weights_log_s = process.recWeights @ np.log(self.samples.s[ind[:process.mu]])
        self.mutProfile.smean = np.exp(weights_log_s) / correctionTerm
        # Safety clamp for global step size
        if not np.isfinite(self.mutProfile.smean) or self.mutProfile.smean < 1e-12:
            self.mutProfile.smean = 1e-12
        
        # Update the covariance matrix
        suggC = np.zeros((problem.dim, problem.dim))
        for parNo in range(process.mu):
            Z = (self.samples.X[ind[parNo], :] - oldCenter) / self.samples.s[ind[parNo]]
            suggC += process.recWeights[parNo] * np.outer(Z, Z)
        
        cc = 1 / process.coreSpecStrPar.tauCov
        newC = (1 - cc) * self.mutProfile.C + cc * suggC
        # Enforce symmetry explicitly and sanitize any numerical issues
        newC = 0.5 * (newC + newC.T)
        # Replace inf/NaN with finite values (fallback to previous covariance)
        if not np.isfinite(newC).all():
            newC = np.where(np.isfinite(newC), newC, self.mutProfile.C)
        # Add a tiny jitter to the diagonal to avoid singularities
        jitter = 1e-12 * np.eye(problem.dim)
        self.mutProfile.C = newC + jitter

        # Perform stable eigen decomposition (matrix is symmetric)
        try:
            eigvals, eigvecs = np.linalg.eigh(self.mutProfile.C)
        except np.linalg.LinAlgError:
            # Fallback: use identity
            eigvals = np.ones(problem.dim)
            eigvecs = np.eye(problem.dim)
            self.mutProfile.C = np.eye(problem.dim)
        # Clamp eigenvalues to a minimum positive threshold
        eigvals = np.maximum(eigvals, 1e-20)
        self.mutProfile.rotMat = eigvecs
        self.mutProfile.stretch = np.sqrt(eigvals)
        
        # Calculate inverse of covariance matrix
        self.mutProfile.Cinv = self.mutProfile.rotMat @ np.diag(self.mutProfile.stretch ** (-2)) @ self.mutProfile.rotMat.T
        
        # Update elite solutions
        if process.coreSpecStrPar.numElt > 0:
            surviveInd = np.argsort(self.samples.f)
            # Count feasible solutions (ignoring NaN values)
            num_feasible = np.sum(self.samples.isFeas[~np.isnan(self.samples.isFeas)])
            limit1 = num_feasible + len(self.elite.val)
            limit2 = process.coreSpecStrPar.numElt
            actEltNum = int(0.5 + min(limit1, limit2))
            
            self.elite.sol = self.samples.X[surviveInd[:actEltNum], :]
            self.elite.val = self.samples.f[surviveInd[:actEltNum]]
            self.elite.Z = self.samples.Z[surviveInd[:actEltNum], :]
            self.elite.s = self.samples.s[surviveInd[:actEltNum]]
            self.elite.wasRepaired = self.samples.wasRepaired[surviveInd[:actEltNum]]
    
    def select(self, restart, archive, process, opt, problem):
        """Perform selection."""
        # Indexes of best non-elite samples
        self.samples.argsortNoElite = np.argsort(self.samples.f)
        
        # Append acceptable elite solutions to the selection pool
        if opt.coreSearch.algorithm == 'CMSA' and process.coreSpecStrPar.numElt > 0 and len(self.elite.val) > 0:
            appendIt = np.ones(len(self.elite.val), dtype=bool)
            
            # Only append elite solutions that are taboo acceptable
            for eltNo in range(len(self.elite.val)):
                appendIt[eltNo] = self.is_taboo_acceptable(self.elite.sol[eltNo, :], 1, opt)
            
            appendIdx = np.where(appendIt)[0]
            
            # Append the approved elite solutions to the selection pool
            self.samples.X = np.vstack([self.samples.X, self.elite.sol[appendIdx, :]])
            self.samples.Z = np.vstack([self.samples.Z, self.elite.Z[appendIdx, :]])
            self.samples.s = np.concatenate([self.samples.s, self.elite.s[appendIdx]])
            self.samples.f = np.concatenate([self.samples.f, self.elite.val[appendIdx]])
            self.samples.wasRepaired = np.concatenate([self.samples.wasRepaired, self.elite.wasRepaired[appendIdx]])
        
        # For performing elite selection
        self.samples.argsortWithElite = np.argsort(self.samples.f)
